ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs

```diff
@@ -261,88 +258,52 @@ applyHelper ::
-    -- The genesis delegates can change within an epoch, so we must check them
-    -- each time.
-    ledgerViewChanged :: SL.ShelleyState c -> SL.ShelleyState c -> Bool
-    ledgerViewChanged old new =
-           SL.nesEL old /= SL.nesEL new
-        || SL.lvGenDelegs oldLedgerView /= SL.lvGenDelegs newLedgerView
-      where
-        oldLedgerView = SL.currentLedgerView old
-        newLedgerView = SL.currentLedgerView new
-
 instance TPraosCrypto c => LedgerSupportsProtocol (ShelleyBlock c) where
   protocolLedgerView _cfg = TickedPraosLedgerView
                           . SL.currentLedgerView
                           . tickedShelleyState
 
-  ledgerViewForecastAt cfg ledgerState at = do
-      guard (at >= minLo)
-      return $ Forecast at $ \for ->
-        case History.find (NotOrigin for) history of
-          Just lv -> return (TickedPraosLedgerView lv)
-          Nothing -> do
-            when (for >= maxHi) $
-              throwError $ OutsideForecastRange {
-                  outsideForecastAt     = at
-                , outsideForecastMaxFor = maxHi
-                , outsideForecastFor    = for
-                }
-            -- 'futureLedgerView' imposes its own bounds, but those bounds are
-            -- set assuming that we are looking forward from the " current "
-            -- ledger state ('shelleyState'), not from the intersection point
-            -- ('at'). Those bounds could /exceed/ the 'maxHi' we have computed,
-            -- but should never be /less/.
-            return $ either (error "futureLedgerView failed") TickedPraosLedgerView $
-                       SL.futureLedgerView globals shelleyState for
+  ledgerViewForecastAt cfg ledgerState = Forecast at $ \for -> if
+      | NotOrigin for == at ->
+        return $ TickedPraosLedgerView $ SL.currentLedgerView shelleyState
+      | for < maxFor ->
+        return $ futureLedgerView for
+      | otherwise ->
+        throwError $ OutsideForecastRange {
+            outsideForecastAt     = at
+          , outsideForecastMaxFor = maxFor
+          , outsideForecastFor    = for
+          }
     where
-      ShelleyLedgerState {history , shelleyState} = ledgerState
+      ShelleyLedgerState { shelleyState } = ledgerState
       globals = shelleyLedgerGlobals cfg
       swindow = SL.stabilityWindow globals
-      tip     = ledgerTipSlot ledgerState
+      at      = ledgerTipSlot ledgerState
 
-      -- Inclusive lower bound
-      minLo :: WithOrigin SlotNo
-      minLo = case tip of
-                NotOrigin (SlotNo s) | s >= swindow -> NotOrigin (SlotNo (s - swindow))
-                _otherwise                          -> Origin
+      -- | 'SL.futureLedgerView' imposes its own bounds. Those bounds could
+      -- /exceed/ the 'maxFor' we have computed, but should never be /less/.
+      futureLedgerView :: SlotNo -> Ticked (SL.LedgerView c)
+      futureLedgerView =
+            either
+              (\e -> error ("futureLedgerView failed: " <> show e))
```

Comment:
Independent: let `futureLedgerView` use reapply instead of apply, then it's no longer partial.